libbeat/beat/event.go

```diff
@@ -173,90 +197,150 @@ func (e *Event) deepUpdate(d mapstr.M, overwrite bool) {
 		e.Fields = mapstr.M{}
 	}
 
-	if overwrite {
+	switch mode {
+	case updateModeOverwrite:
 		e.Fields.DeepUpdate(d)
-	} else {
+	case updateModeNoOverwrite:
 		e.Fields.DeepUpdateNoOverwrite(d)
 	}
 }
 
-func (e *Event) setTimestamp(v interface{}) error {
+func (e *Event) setTimestamp(v interface{}) (interface{}, error) {
+	// to satisfy the PutValue interface, this function
+	// must return the overwritten value
+	prevValue := e.Timestamp
+
 	switch ts := v.(type) {
 	case time.Time:
 		e.Timestamp = ts
+		return prevValue, nil
 	case common.Time:
 		e.Timestamp = time.Time(ts)
+		return prevValue, nil
 	default:
-		return errNoTimestamp
+		return nil, ErrValueNotTimestamp
 	}
-
-	return nil
 }
 
+// Put associates the specified value with the specified key. If the event
+// previously contained a mapping for the key, the old value is replaced and
+// returned. The key can be expressed in dot-notation (e.g. x.y) to put a value
+// into a nested map.
+//
+// If you need insert keys containing dots then you must use bracket notation
+// to insert values (e.g. m[key] = value).
```

Comment:
```suggestion
// If you need to insert keys containing dots then you must use bracket notation
// to insert values (e.g. m[key] = value).
```

Reply:
It's a copy-paste from here https://github.com/elastic/elastic-agent-libs/blob/0e722a9ae727502792aa7dbeccf3a78bc987e5a6/mapstr/mapstr.go#L193

I'll fix this typo when I have a chance.